052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,156 @@
+
+
+## Proposal
+
+The proposal is to build a metrics reporter, `PrometheusMetricsReporter`, that implements both `org.apache.kafka.common.metrics.MetricsReporter` and `kafka.metrics.KafkaMetricsReporter`, and that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/052-proposal.png)
+
+To enable the reporter, the following Kafka configurations must be set:
+
+```
+metric.reporters=com.mickaelmaison.PrometheusMetricsReporter # For KafkaMetrics
+kafka.metrics.reporters=com.mickaelmaison.PrometheusMetricsReporter # For YammerMetrics - This is only required for brokers/controllers.
+```
+
+The reporter will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it now or later to the reporter.
+Today ZooKeeper also uses jmx_exporter to expose its metrics. ZooKeeper is going away and will be replaced by KRaft controllers which also support Kafka metrics reporters.
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Since ZooKeeper does not support metrics reporters, we should keep the current mechanism using jmx_exporter while we support ZooKeeper deployments.
+
+The reporter will be a new project and should be usable without Strimzi for both brokers and client side applications. As it's new code, it should be first release as a preview feature before becoming the preferred mechanism once KRaft is the default.
+
+In `strimzi-kafka-operator`, we will need new examples demonstrating how to use the reporter, new grafana dashboards with metric names from the reporter.
+
+## Compatibility
+
+Differences with jmx_exporter metrics:
+
+- The reporter does not compute 1/5/15 minute rate, mean, max, min, stddev metrics. It's preferable to compute them in Prometheus instead of on the client side.
+- The reporter is missing the `kafka_server_app_info_starttimems` metric with the client/broker id label. (Due to [KAFKA-15186](https://issues.apache.org/jira/browse/KAFKA-15186))
+
+Assuming jmx_exporter does not have any rules, this is the other main metric change:
+
+- With the reporter, the `name` field is put directly in the metric name. For example this MBean, `kafka.server:type=ZooKeeperClientMetrics,name=ZooKeeperRequestLatencyMs` with the `Count` attribute is converted to `kafka_server_zookeeperclientmetrics_zookeeperrequestlatencyms_count`. By default jmx_exporter keeps the name as a label, `kafka_server_zookeeperclientmetrics_count{name="ZooKeeperRequestLatencyMs",}`.
```

Comment:
This would imply that existing dashboards won't be compatible, right? Perhaps, in addition to Jakub's comment above, you can elaborate on how the example dashboards would be migrated.

Reply:
Have you evaluated if the default/current configuration we provide for JMX exporter could be a viable way to set the naming for the new exporter in order to keep compatibility? So in the example above why not having `name` as a label instead of part of the metric name?

Reply:
This section only applies if jmx_exporter is used without any rules. As explained in the paragraph below, with the default rules we (and the jmx_exporter project) provide, jmx_exporter generates metric names that are very similar to the ones generated by the reporter.